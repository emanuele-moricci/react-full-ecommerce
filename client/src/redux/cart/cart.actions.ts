import { Item } from "src/redux/shop/shop.types";
import { CartActionTypes } from "src/redux/cart/cart.types";

export const toggleCartDropdown = () => ({
  type: CartActionTypes.TOGGLE_CART_DROPDOWN,
});

export const addItem = (item: Item) => ({
  type: CartActionTypes.ADD_ITEM,
  payload: item,
});

export const removeItem = (item: Item) => ({
  type: CartActionTypes.REMOVE_ITEM,
  payload: item,
});

export const deleteItemFromCart = (item: Item) => ({
  type: CartActionTypes.DELETE_ITEM_FROM_CART,
  payload: item,
});

export const clearCart = () => ({
  type: CartActionTypes.CLEAR_CART,
});

export const updateCartOnDatabase = () => ({
  type: CartActionTypes.UPDATE_CART_IN_DATABASE,
});

export const setCartFromDatabase = (cartItems: Item[]) => ({
  type: CartActionTypes.SET_CART_FROM_DATABASE,
  payload: cartItems,
});
