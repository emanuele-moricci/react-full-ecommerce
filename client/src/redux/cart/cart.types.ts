import { Item } from "src/pages/shop/state/shop.types";

export type CartItem = Item & {
  quantity: number;
};
