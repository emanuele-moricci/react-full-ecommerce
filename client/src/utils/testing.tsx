import { FC } from "react";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import { BrowserRouter } from "react-router-dom";

export const fakePayloadAction = { type: "FAKE_PAYLOAD" };

export const getMockedStoreRenderedComponent = (state: any, Component: FC) => {
  const mockStore = configureStore();
  const store = mockStore(state);

  return renderer.create(
    <Provider store={store}>
      <BrowserRouter>
        <Component />
      </BrowserRouter>
    </Provider>
  );
};
