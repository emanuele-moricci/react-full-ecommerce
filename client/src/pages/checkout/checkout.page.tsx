import { useEffect } from "react";
import { useNavigate } from "react-router";
import { useSelector } from "react-redux";
import { selectItems, selectCartTotal } from "src/redux/cart/cart.selectors";
import { selectCheckoutPurchasing } from "src/redux/checkout/checkout.selectors";

import CheckoutItem from "src/components/checkout/checkout-item/checkout-item.component";
import StripeCheckoutButton from "src/components/checkout/stripe-button/stripe-button.component";
import Spinner from "src/components/layout/spinner/spinner.component";

import * as Styled from "./checkout.styles";

const Checkout = (): JSX.Element => {
  const navigate = useNavigate();

  const items = useSelector(selectItems);
  const total = useSelector(selectCartTotal);
  const purchasing = useSelector(selectCheckoutPurchasing);

  useEffect(() => {
    if (!items.length) navigate("/");
  }, [items, navigate]);

  return purchasing ? (
    <Spinner />
  ) : (
    <Styled.CheckoutPage>
      <Styled.CheckoutHeader>
        <Styled.HeaderBlock>
          <span>Product</span>
        </Styled.HeaderBlock>
        <Styled.HeaderBlock>
          <span>Description</span>
        </Styled.HeaderBlock>
        <Styled.HeaderBlock>
          <span>Quantity</span>
        </Styled.HeaderBlock>
        <Styled.HeaderBlock>
          <span>Price</span>
        </Styled.HeaderBlock>
        <Styled.HeaderBlock>
          <span>Remove</span>
        </Styled.HeaderBlock>
      </Styled.CheckoutHeader>
      {items.map((item) => (
        <CheckoutItem key={item.id} item={item} />
      ))}
      <Styled.Total>
        <span>TOTAL: ${total}</span>
      </Styled.Total>
      <Styled.TestWarning>
        *Please use the following test credit card for payments*
        <br />
        4242 4242 4242 4242 - Exp: 01/29 - CVV: 123
      </Styled.TestWarning>
      <StripeCheckoutButton price={total} />
    </Styled.CheckoutPage>
  );
};

export default Checkout;
