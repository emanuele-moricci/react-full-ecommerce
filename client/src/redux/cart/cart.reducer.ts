import { AnyAction } from "redux";
import { CartActionTypes } from "src/redux/cart/cart.types";
import {
  addItem,
  removeItem,
  deleteItemFromCart,
} from "src/redux/cart/cart.utils";

const INITIAL_STATE = {
  hidden: true,
  items: [],
};

const cartReducer = (state = INITIAL_STATE, { type, payload }: AnyAction) => {
  switch (type) {
    case CartActionTypes.TOGGLE_CART_DROPDOWN:
      return {
        ...state,
        hidden: !state.hidden,
      };
    case CartActionTypes.ADD_ITEM:
      return {
        ...state,
        items: addItem(state.items, payload),
      };
    case CartActionTypes.REMOVE_ITEM:
      return {
        ...state,
        items: removeItem(state.items, payload),
      };
    case CartActionTypes.DELETE_ITEM_FROM_CART:
      return {
        ...state,
        items: deleteItemFromCart(state.items, payload),
      };
    case CartActionTypes.CLEAR_CART:
      return {
        ...state,
        items: [],
      };
    case CartActionTypes.SET_CART_FROM_DATABASE:
      return {
        ...state,
        items: payload,
      };
    default:
      return state;
  }
};

export default cartReducer;
