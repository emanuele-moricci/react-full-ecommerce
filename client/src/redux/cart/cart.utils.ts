import { Item } from "src/redux/shop/shop.types";
import { CartItem } from "src/redux/cart/cart.types";

export const addItem = (
  cartItems: CartItem[],
  cartItemToAdd: Item
): CartItem[] => {
  const existingCartItem = cartItems.find(
    (cartItem) => cartItem.id === cartItemToAdd.id
  );

  if (existingCartItem) {
    return cartItems.map((cartItem) =>
      cartItem.id === cartItemToAdd.id
        ? { ...cartItem, quantity: cartItem.quantity + 1 }
        : cartItem
    );
  }

  return [...cartItems, { ...cartItemToAdd, quantity: 1 }];
};

export const removeItem = (
  cartItems: CartItem[],
  cartItemToRemove: Item
): CartItem[] => {
  const existingCartItem = cartItems.find(
    (cartItem) => cartItem.id === cartItemToRemove.id
  );

  if (existingCartItem?.quantity === 1) {
    return deleteItemFromCart(cartItems, cartItemToRemove);
  }

  return cartItems.map((cartItem) =>
    cartItem.id === cartItemToRemove.id
      ? { ...cartItem, quantity: cartItem.quantity - 1 }
      : cartItem
  );
};

export const deleteItemFromCart = (
  cartItems: CartItem[],
  cartItemToDelete: Item
): CartItem[] => {
  return cartItems.filter((cartItems) => cartItems.id !== cartItemToDelete.id);
};
