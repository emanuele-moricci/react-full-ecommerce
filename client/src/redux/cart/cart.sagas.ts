import { PayloadAction } from "@reduxjs/toolkit";
import {
  all,
  call,
  takeLatest,
  put,
  PutEffect,
  select,
} from "redux-saga/effects";

import { cartActions } from "./cart.slice";
import { userActions } from "src/redux/user/user.slice";
import { Item } from "src/pages/shop/state/shop.types";
import { User } from "src/redux/user/user.types";
import { selectItems } from "./cart.selectors";
import { selectCurrentUser } from "src/redux/user/user.selectors";

import { getUserCartRef } from "src/db/firebase.utils";
import {
  DocumentData,
  DocumentReference,
  DocumentSnapshot,
  getDoc,
  updateDoc,
} from "@firebase/firestore";

/* ∨∨∨∨ START FUNCTIONS ∨∨∨∨ */

export function* cartSagas() {
  yield all([
    call(onSignInSuccess),
    call(onSignOutSuccess),
    call(onCartChange),
  ]);
}

export function* onSignInSuccess() {
  yield takeLatest(userActions.authSuccess.type, checkCartFromDatabase);
}

export function* onSignOutSuccess() {
  yield takeLatest(userActions.SignOutSuccess.type, clearCartOnSignOut);
}

export function* onCartChange() {
  yield takeLatest(
    [
      cartActions.addItem.type,
      cartActions.removeItem.type,
      cartActions.deleteItemFromCart.type,
      cartActions.updateCartOnDatabase.type,
    ],
    updateCartOnDatabase
  );
}

/* ^^^^ START FUNCTIONS ^^^^ */

function* clearCartOnSignOut(): Generator<PutEffect> {
  yield put(cartActions.clearCart());
}

export function* checkCartFromDatabase({
  payload: { id },
}: PayloadAction<{ id: string }>): Generator {
  const cartRef = (yield getUserCartRef(id)) as DocumentReference<DocumentData>;
  const cartSnapshot = (yield getDoc(
    cartRef
  )) as DocumentSnapshot<DocumentData>;
  yield put(cartActions.setCartFromDatabase(cartSnapshot.data()?.cartItems));
}

function* updateCartOnDatabase(): Generator {
  const currentUser = (yield select(selectCurrentUser)) as User;
  if (currentUser) {
    try {
      const cartRef = (yield getUserCartRef(
        currentUser.id
      )) as DocumentReference<DocumentData>;
      const cartItems = (yield select(selectItems)) as Item[];

      yield updateDoc(cartRef, { cartItems });
    } catch (error) {
      console.log(error);
    }
  }
}
