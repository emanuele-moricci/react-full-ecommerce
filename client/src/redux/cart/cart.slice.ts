import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { CartItem } from "./cart.types";
import { Item } from "src/pages/shop/state/shop.types";
import { addItem, deleteItemFromCart, removeItem } from "./cart.utils";

type InitialStateType = {
  hidden: boolean;
  items: CartItem[];
};

const initialState: InitialStateType = {
  hidden: true,
  items: [],
};

const slice = createSlice({
  name: "cart",
  initialState,
  reducers: {
    toggleCartDropdown: (state) => {
      state.hidden = !state.hidden;
    },
    addItem: (state, { payload }: PayloadAction<Item>) => {
      state.items = addItem(state.items, payload);
    },
    removeItem: (state, { payload }: PayloadAction<Item>) => {
      state.items = removeItem(state.items, payload);
    },
    deleteItemFromCart: (state, { payload }: PayloadAction<Item>) => {
      state.items = deleteItemFromCart(state.items, payload);
    },
    clearCart: (state) => {
      state.items = [];
    },
    setCartFromDatabase: (state, { payload }: PayloadAction<CartItem[]>) => {
      state.items = payload;
    },
    updateCartOnDatabase: () => {},
  },
});

export const cartActions = slice.actions;
export default slice.reducer;
