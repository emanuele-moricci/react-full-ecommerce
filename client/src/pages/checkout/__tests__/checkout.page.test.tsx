import { getMockedStoreRenderedComponent } from "src/utils/testing";

import { initialState as checkoutState } from "../state/checkout.slice";
import { initialState as cartState } from "src/redux/cart/cart.slice";

import Checkout from "src/pages/checkout/checkout.page";

describe("<Checkout />", () => {
  it("Should match the snapshot", () => {
    const component = getMockedStoreRenderedComponent(
      { checkout: checkoutState, cart: cartState },
      Checkout
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
