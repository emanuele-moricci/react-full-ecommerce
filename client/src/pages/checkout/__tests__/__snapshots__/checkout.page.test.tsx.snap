// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkout /> Should match the snapshot 1`] = `
<div
  className="sc-jrQzAO coUdoP"
>
  <div
    className="sc-kDTinF jFesrt"
  >
    <div
      className="sc-iqseJM dfxpsH"
    >
      <span>
        Product
      </span>
    </div>
    <div
      className="sc-iqseJM dfxpsH"
    >
      <span>
        Description
      </span>
    </div>
    <div
      className="sc-iqseJM dfxpsH"
    >
      <span>
        Quantity
      </span>
    </div>
    <div
      className="sc-iqseJM dfxpsH"
    >
      <span>
        Price
      </span>
    </div>
    <div
      className="sc-iqseJM dfxpsH"
    >
      <span>
        Remove
      </span>
    </div>
  </div>
  <div
    className="sc-crHmcD dzxRkv"
  >
    <span>
      TOTAL: $
      0
    </span>
  </div>
  <div
    className="sc-egiyK eyBjwp"
  >
    *Please use the following test credit card for payments*
    <br />
    4242 4242 4242 4242 - Exp: 01/29 - CVV: 123
  </div>
  <button
    className="StripeCheckout"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onMouseDown={[Function]}
    onMouseOut={[Function]}
    onMouseUp={[Function]}
    style={
      Object {
        "background": "linear-gradient(#28a0e5,#015e94)",
        "border": 0,
        "borderRadius": 5,
        "boxShadow": "0 1px 0 rgba(0,0,0,0.2)",
        "cursor": "pointer",
        "display": "inline-block",
        "overflow": "hidden",
        "padding": 1,
        "textDecoration": "none",
        "userSelect": "none",
        "visibility": "visible",
      }
    }
  >
    <span
      style={
        Object {
          "backgroundImage": "linear-gradient(#7dc5ee,#008cdd 85%,#30a2e4)",
          "borderRadius": 4,
          "boxShadow": "inset 0 1px 0 rgba(255,255,255,0.25)",
          "color": "#fff",
          "display": "block",
          "fontFamily": "\\"Helvetica Neue\\",Helvetica,Arial,sans-serif",
          "fontSize": 14,
          "fontWeight": "bold",
          "height": 30,
          "lineHeight": "30px",
          "padding": "0 12px",
          "position": "relative",
          "textShadow": "0 -1px 0 rgba(0,0,0,0.25)",
        }
      }
    >
      Pay Now
    </span>
  </button>
</div>
`;
